### ISSUE TITLE
Make error more descriptive when non-sink argument is used instead of expected sink argument

### ISSUE THREAD

**Issue by Andrew15-5** [typst/typst#2102]
The issue #2101 was created due to the next error message:

```
error: unexpected argument
   ┌─ /file.typ:42:21
   │
42 │   heading(numbering: it => it, name, ..args)
   │                      ^^
```

It shows that the argument is unexpected, but in reality it is of course expected, and instead of a normal argument, it should be a [sink argument](https://typst.app/docs/reference/types/arguments/):

```typ
42 │   heading(numbering: (..it) => it, name, ..args)
```

The error did throw me off, but after an [answer](https://github.com/typst/typst/issues/2101#issuecomment-1712825858) and some "digging" through the [documentation](https://github.com/typst/typst/issues/2101#issuecomment-1712842441), the error was resolved. In any case, the error message should be more descriptive and user-friendly.

**Comment by tingerrr** [comment]
So naively, this would be applied to the result of `Numbering::{apply_vm,apply_vt}` if any of the contained errors was of this type, but errors currently don't carry any strongly typed kind.

I was actually thinking of adding these when I wrote #2385, but refrained as it was out of scope and had not immediate benefit.

Personally, I think adding an error code to diagnostics would make attaching hints to diagnostics much easier, as we wouldn't need to rely on string comparisons to find out if we have a specific diagnostic category here.

**Comment by Yoric** [comment]
Is this issue still valid? If so, I'd be happy to take a look at it.

**Comment by laurmaedje** [comment]
Yes, it is! The confusing situation always arises when the "unexpected argument" error (which is supposed to be annotated where the user typed argument) is instead annotated at the parameter because the argument is compiler generated and has no source location (= span). In this case, the argument's span defaults to the function's span:

https://github.com/typst/typst/blob/9b09146a6b5e936966ed7ee73bce9dd2df3810ae/crates/typst-library/src/foundations/func.rs#L297

The error is then produced here: https://github.com/typst/typst/blob/9b09146a6b5e936966ed7ee73bce9dd2df3810ae/crates/typst-eval/src/call.rs#L250

You could look into detecting that the argument's span matches the function's span (`args.span == func.span()`) and producing a better error message in this case. This would be a simple way to detect the situation. There might be a way to make it cleaner by not assigning the default span in this way at all, but I'm not sure it would work out easily.

**Comment by Yoric** [comment]
Thanks for the reply, @laurmaedje. Trying to parse this.

If I understand correctly your links, the check is performed at the call-site. In the case of

```typst
heading(numbering: it => it, name, ..args)
```

does this mean that the callee is `heading` itself?

**Comment by laurmaedje** [comment]
The check is performed at the call-site of `it => it`, which is within the heading show rule (indirectly through the numbering code):

https://github.com/typst/typst/blob/9b09146a6b5e936966ed7ee73bce9dd2df3810ae/crates/typst-library/src/model/heading.rs#L236-L238

https://github.com/typst/typst/blob/9b09146a6b5e936966ed7ee73bce9dd2df3810ae/crates/typst-library/src/introspection/counter.rs#L671

https://github.com/typst/typst/blob/9b09146a6b5e936966ed7ee73bce9dd2df3810ae/crates/typst-library/src/model/numbering.rs#L112

https://github.com/typst/typst/blob/9b09146a6b5e936966ed7ee73bce9dd2df3810ae/crates/typst-library/src/foundations/func.rs#L297

> To reproduce the error, would it be sufficient to add e.g.

Yes

**Comment by Yoric** [comment]
Ah, the check is perform at the real call-site, but the span points to argument `it` of the anonymous function passed as `numbering`?

**Comment by Yoric** [comment]
So ideally, what kind of error would we want to display? Something like `this function receives a sink argument but expects a non-sink argument`?

**Comment by Andrew15-5** [comment]
Something like that. Plus a hint that says that you gotta try using something like `(..n)`.

**Comment by laurmaedje** [comment]
> So ideally, what kind of error would we want to display? Something like this function receives a sink argument but expects a non-sink argument?

I think that error message isn't too clear, to be honest. The term "sink argument" isn't really a thing. If anything, it would be spreaded arguments. But we don't know whether arguments were spreaded at the point where I proposed the check. We only know that there are multiple arguments and that our anonymous function was not prepared to handle all of them.

> Plus a hint that says that you gotta try using something like (..n).

We also don't know whether an argument sink is the best solution. The function could just always be passed two arguments, but have just one parameter.

---

What the error message should express, in my opinion, is that the function was called with a different number of arguments than it has parameters and that one should either adjust the number of parameters or consider using an argument sink.

**Comment by Andrew15-5** [comment]
> The term "sink argument" isn't really a thing.

It's literally a thing: https://typst.app/docs/reference/foundations/arguments/#argument-sinks.

[Spread operator](https://typst.app/docs/reference/foundations/arguments/#argument-sinks) is when you pass something in, not when you receive something.

> We also don't know whether an argument sink is the best solution.

It is the most idiomatic solution with regards to numbering. It is the solution that 100% will work, because it captures any arguments. Any other solution are more specific and hint can't cover them all, it's just a hint to fix the error.

> The function could just always be passed two arguments, but have just one parameter.

`#let f(one) = {};#f(1, 2)` produces an error, so this statement is incorrect. I don't understand it at all.

> What the error message should express, in my opinion, is that the function was called with a different number of arguments than it has parameters and that one should either adjust the number of parameters or consider using an argument sink.

If it can't detect the callee function/field, then sure, but if it can, then it's better to tailor the explanation to fit better to the situation.

**Comment by tingerrr** [comment]
> > The term "sink argument" isn't really a thing.
> 
> It's literally a thing: https://typst.app/docs/reference/foundations/arguments/#argument-sinks.

Not to be pedantic, but it technically isn't as that link refers to argument sinks, not sink arguments, this is because a sink may receive many arguments.

> [Spread operator](https://typst.app/docs/reference/foundations/arguments/#argument-sinks) is when you pass something in, not when you receive something.
> 
> > We also don't know whether an argument sink is the best solution.
> 
> It is the most idiomatic solution with regards to numbering. It is the solution that 100% will work, because it captures any arguments. Any other solution are more specific and hint can't cover them all, it's just a hint to fix the error.

The problem is that from within Typst, it's not immediately clear if this is related to numbering _at all_. This kind of error can be emitted anywhere where an unexpected number of arguments is passed to some anonymous function.

> > The function could just always be passed two arguments, but have just one parameter.
> 
> `#let f(one) = {};#f(1, 2)` produces an error, so this statement is incorrect. I don't understand it at all.

He's saying that this same error would be emitted in this exact case or similar cases.

> > What the error message should express, in my opinion, is that the function was called with a different number of arguments than it has parameters and that one should either adjust the number of parameters or consider using an argument sink.
> 
> If it can't detect the callee function/field, then sure, but if it can, then it's better to tailor the explanation to fit better to the situation.

This once again assumes that numbering is _the_ titular situation, but it is only one situation where this happens, plenty of packages use callbacks with specific parameter counts, all of which would suffer UX wise if the message was hand tailored to assume it's a numbering function.

**Comment by Yoric** [comment]
I think I'm getting lost in this conversation :)

If I read things correctly, an easy (but probably incomplete) fix would be to simply replace

```rust
bail!(arg.span, "unexpected argument")
```

with something along the lines of

```rust
bail!(arg.span, "wrong number number of arguments, {n} arguments were not consumed", n=self.items.len())
```

without attempting to be smart (i.e. still just two cases). Is this correct?

Now, if I read things correctly, to do things better, we'd need the ability to detect why an argument was not consumed, as it could be because we're passing/expecting a spread, because this would help us customize the error message? At this case, my understanding of the goal is insufficient. If we want to head in this direction, I'd need to have a few examples of the various cases in which we want to improve the error messages.

**Comment by Yoric** [comment]
ping @Andrew15-5  @reknih @tingerrr ?

**Comment by tingerrr** [comment]
I think what you wrote is already an improvement over what we have now. And yes, We need more information to be able to tell the user why an argument may be unexpected. I've briefly looked at this in the past, but at least with my understanding at that time it looked like a non-trivial refactor of the argument handling code.

**Comment by Andrew15-5** [comment]
> ```rs
> bail!(arg.span, "wrong number number of arguments, {n} arguments were not consumed", n=self.items.len())
> ```

[number number](https://www.youtube.com/watch?v=MPd3_nUnKK8)

I think adding number to the error is great. Though I think it's better to say how many arguments were given and used, because this is information that you probably have, not the difference between the two.

I don't know if infrastructure allows detecting numbering case or how hard it will be to make it work. But that would be ideal.

**Comment by Yoric** [comment]
I've been trying to add some more information to `Args` to track the initial number of arguments, but this breaks `test_value_size` by making `Args` too large.

My hope was that the following error message would make sense:

```rust
        if list.len() < n {
            bail!(
                self.span,
                "wrong number number of arguments: {} arguments were given, but we're trying to consume at least {} arguments",
                self.initial_len,
                self.items.len() + n,
            )
        }
```

How should I approach this? I suppose that I could box `Args` in `Value`, but I expect that this will have a performance/memory cost.

**Comment by laurmaedje** [comment]
I think this is derailing a bit. I wouldn't want to replace the "unexpected argument" error everywhere. The message above is a bit verbose and technical in my opinion.

I think we should go back to what I mentioned in https://github.com/typst/typst/issues/2102#issuecomment-2858747107 and only have a more specific message for the cases where both spans match, e.g. something in the direction of `function received more arguments than it expected`.
